### PR TITLE
[camerax] Make fixes required to swap camera_android_camerax for camera_android

### PR DESCRIPTION
**Pull request by camsim99**
Makes changes needed to land https://github.com/flutter/packages/pull/6629. Specifically:

- Fixes timing issue with `stopVideoRecording` such that the `Future` it returns will only complete when CameraX reports that the recording is finalized (via listening for the [finalized video recording event](https://developer.android.com/reference/androidx/camera/video/VideoRecordEvent.Finalize))
- Modifies `startVideoCapturing` such that the `Future` it returns will only complete when CameraX reports that video capturing has started (via listening for the [started video recording event](https://developer.android.com/reference/androidx/camera/video/VideoRecordEvent.Start))
- Adds empty implementation and TODO for implementing `setDescriptionWhileRecording`

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart [hunk 1]
```diff
@@ -122,7 +128,7 @@ class AndroidCameraCameraX extends CameraPlatform {
 
   /// The prefix used to create the filename for video recording files.
   @visibleForTesting
-  final String videoPrefix = 'MOV';
+  final String videoPrefix = 'REC';
```
Comment (camsim99) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:131]: Changed this + the .temp --> .mp4 extension change to be consistent with camera_android, see [source](https://github.com/flutter/packages/blob/d670b2c38c8db0a773aa703e7d3f15682e05ad7f/packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java#L1250).

File: packages/camera/camera_android_camerax/README.md [hunk 2]
```diff
@@ -64,11 +67,4 @@ For more information on contributing to this plugin, see [`CONTRIBUTING.md`](CON
 [6]: https://developer.android.com/media/camera/camerax/architecture#combine-use-cases
 [7]: https://developer.android.com/reference/android/hardware/camera2/CameraMetadata#INFO_SUPPORTED_HARDWARE_LEVEL_3
 [8]: https://developer.android.com/reference/android/hardware/camera2/CameraMetadata#INFO_SUPPORTED_HARDWARE_LEVEL_LIMITED
-[120462]: https://github.com/flutter/flutter/issues/120462
```
Comment (reidbaker) [inline packages/camera/camera_android_camerax/README.md:67]: Intentional removal?

File: packages/camera/camera_android_camerax/example/integration_test/integration_test.dart [hunk 3]
```diff
@@ -178,4 +179,86 @@ void main() {
...
+    await controller.prepareForVideoRecording();
+
+    await controller.startVideoRecording();
+    final int recordingStart = DateTime.now().millisecondsSinceEpoch;
+
+    sleep(const Duration(seconds: 2));
+
+    final XFile file = await controller.stopVideoRecording();
+    final int recordingTime =
+        DateTime.now().millisecondsSinceEpoch - recordingStart;
```
Comment (reidbaker) [inline packages/camera/camera_android_camerax/example/integration_test/integration_test.dart:202]: I think we can make this test a bit better by comparing on both sides of the stopRecording call. 


```suggestion
    // Test should pass with this at any non zero value. 
    sleep(const Duration(seconds: 2));
    
     final int preStopTime =
        DateTime.now().millisecondsSinceEpoch - recordingStart;
    final XFile file = await controller.stopVideoRecording();
    final int postStopTime =
        DateTime.now().millisecondsSinceEpoch - recordingStart;
        ....
        expect(duration, greaterThan(preStopTime));
        expect(duration, lessThan(postStopTime));
```

Note: 
We should be using a package like https://pub.dev/packages/system_clock because DateTime.now is vulnerable to the clock changing while the test is running but that feels out of scope.

File: packages/camera/camera_android_camerax/example/integration_test/integration_test.dart [hunk 4]
```diff
@@ -178,4 +179,86 @@ void main() {
...
+    await controller.prepareForVideoRecording();
+
+    int startPause;
+    int timePaused = 0;
+
+    await controller.startVideoRecording();
+    final int recordingStart = DateTime.now().millisecondsSinceEpoch;
+    sleep(const Duration(milliseconds: 500));
+
+    await controller.pauseVideoRecording();
```
Comment (reidbaker) [inline packages/camera/camera_android_camerax/example/integration_test/integration_test.dart:234]: Possibly overkill but consider dropping the pause/"measure pause time" into either a loop or a method. When we were debugging having the blocks back to back made it harder to understand until you realized the blocks where supposed to be identical.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart (hunk 1 above)
Comment (reidbaker) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:131]: Are we guaranteed that the file is an mp4? If not then maybe we should leave it .temp because having the wrong file extension is worse than having an unknown file extension.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart [hunk 5]
```diff
@@ -777,6 +783,15 @@ class AndroidCameraCameraX extends CameraPlatform {
     await _unbindUseCaseFromLifecycle(preview!);
   }
 
+  /// Sets the active camera while recording.
+  ///
+  /// Currently unsupported, so is a no-op.
+  @override
+  Future<void> setDescriptionWhileRecording(CameraDescription description) {
```
Comment (reidbaker) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:790]: Does this mean we have a feature we know will silently stop working when we modify the default to camerax?

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart [hunk 6]
```diff
@@ -979,21 +993,22 @@ class AndroidCameraCameraX extends CameraPlatform {
           'Attempting to stop a '
               'video recording while no recording is in progress.');
     }
+
+    /// Stop the active recording and wiat for the video recording to be finalized.
```
Comment (reidbaker) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:997]:
```suggestion
    /// Stop the active recording and wait for the video recording to be finalized.
```

**Review by reidbaker — APPROVED** [review]

File: packages/camera/camera_android_camerax/README.md (hunk 2 above)
Comment (camsim99) [reply packages/camera/camera_android_camerax/README.md:67]: Yes! There aren't referenced in the README anymore.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart (hunk 5 above)
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:790]: It does mean that :/ but the fact that it will be a breaking change can mitigate that fact; it will be noted in the README here but I can also add it to the camera/camera CHANGELOG and camera/camera README.

@stuartmorgan any thoughts on this? I realized recently that this was not implemented for camera_android_camerax.

Comment (stuartmorgan) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:790]: I would definitely note the loss of functionality in the CHANGELOG for the `camera` breaking change.

This is a pretty new feature IIRC, and I would expect relatively niche, so losing it in a breaking change doesn't seem like a major issue. People will always have the possibility of using `camera_android` (or a fork of it, if/when we stop maintaining it) if they need it.

File: packages/camera/camera_android_camerax/example/integration_test/integration_test.dart (hunk 3 above)
Comment (camsim99) [reply packages/camera/camera_android_camerax/example/integration_test/integration_test.dart]: The preStopTime check didn't quite work I assume because stopping video recording takes so long but I kept the naming of postStopTime for clarity

Comment (camsim99) [reply packages/camera/camera_android_camerax/example/integration_test/integration_test.dart]: I suppose we could also fix this by listening for the [`VideoRecordEvent.Start`](https://developer.android.com/reference/androidx/camera/video/VideoRecordEvent.Start) event to return. Let me give it a go!

Comment (camsim99) [reply packages/camera/camera_android_camerax/example/integration_test/integration_test.dart]: @reidbaker Ok so I added the fix but it didn't quite work still :/ the duration increased but guessing there still some lag time in starting the video. Still left it in though as it is a better approximation.

Comment (camsim99) [reply packages/camera/camera_android_camerax/example/integration_test/integration_test.dart]: Going to land this as-is since I've done all I can to remedy this issue. If we feel like we need this fixed, I can file a bug and escalate to CameraX.

Comment (camsim99) [reply packages/camera/camera_android_camerax/example/integration_test/integration_test.dart]: Filed https://github.com/flutter/flutter/issues/148138 for this.
